src/utils/translator.rs

```diff
@@ -0,0 +1,494 @@
+			_ => return Err(TranslatorError::NoMutationsApplicable),
+		}
+	}
+	t.translate_op(&op)?.encode(&mut offset_bytes);
+	match e.read()? {
+		Operator::End if e.eof() => {},
+		_ => return Err(TranslatorError::NoMutationsApplicable),
+	}
+	Ok(wasm_encoder::ConstExpr::raw(offset_bytes))
+}
+
+#[allow(dead_code)]
+pub fn element(
+	t: &mut dyn Translator,
+	element: wasmparser::Element<'_>,
+	s: &mut ElementSection,
+) -> Result<()> {
+	let offset;
+	let mode = match &element.kind {
+		ElementKind::Active { table_index, offset_expr } => {
+			offset = t.translate_const_expr(
+				offset_expr,
+				&wasmparser::ValType::I32,
+				ConstExprKind::ElementOffset,
+			)?;
+			let table_index = table_index.unwrap_or(0);
+			let table = t.remap(Item::Table, table_index)?;
+			ElementMode::Active {
+				table: if table == 0 { None } else { Some(table) },
+				offset: &offset,
+			}
+		},
+		ElementKind::Passive => ElementMode::Passive,
+		ElementKind::Declared => ElementMode::Declared,
+	};
+	let element_type = t.translate_refty(&element.ty)?;
+	let functions;
+	let exprs;
+	let elements = match element.items {
+		wasmparser::ElementItems::Functions(reader) => {
+			functions = reader
+				.into_iter()
+				.map(|f| t.remap(Item::Function, f?))
+				.collect::<wasmparser::Result<Vec<_>, _>>()?;
+			Elements::Functions(&functions)
+		},
+		wasmparser::ElementItems::Expressions(reader) => {
+			exprs = reader
+				.into_iter()
+				.map(|f| {
+					t.translate_const_expr(
+						&f?,
+						&wasmparser::ValType::Ref(element.ty),
+						ConstExprKind::ElementFunction,
+					)
+				})
+				.collect::<wasmparser::Result<Vec<_>, _>>()?;
+			Elements::Expressions(&exprs)
+		},
+	};
+	s.segment(ElementSegment { mode, element_type, elements });
+	Ok(())
+}
+
+#[allow(unused_variables)]
+pub fn op(t: &mut dyn Translator, op: &Operator<'_>) -> Result<Instruction<'static>> {
+	use wasm_encoder::Instruction as I;
+
+	macro_rules! translate {
+        ($( @$proposal:ident $op:ident $({ $($arg:ident: $argty:ty),* })? => $visit:ident)*) => {
+            Ok(match op {
+                $(
+                    wasmparser::Operator::$op $({ $($arg),* })? => {
+                        $(
+                            $(let $arg = translate!(map $arg $arg);)*
+                        )?
+                        translate!(build $op $($($arg)*)?)
+                    }
+                )*
+            })
+        };
+
+        // This case is used to map, based on the name of the field, from the
+        // wasmparser payload type to the wasm-encoder payload type through
+        // `Translator` as applicable.
+        (map $arg:ident tag_index) => (t.remap(Item::Tag, *$arg)?);
+        (map $arg:ident function_index) => (t.remap(Item::Function, *$arg)?);
+        (map $arg:ident table) => (t.remap(Item::Table, *$arg)?);
+        (map $arg:ident table_index) => (t.remap(Item::Table, *$arg)?);
+        (map $arg:ident table) => (t.remap(Item::Table, *$arg)?);
```

Comment:
I know the first argument to `remap` is ignored, but the code doesn't look right. Both table index and table maps to `Item::Table`.

Reply:
Right. This will be no longer an issue after moving to original version